staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -27,12 +27,54 @@ import (
 
 // DelayingInterface is an Interface that can Add an item at a later time. This makes it easier to
 // requeue items after failures without ending up in a hot-loop.
+// A Delaying queue can be thought of two separate queues that work together - an 'active' queue of
+// items which we want to process ASAP and a 'waiting' queue of items which must wait for a period of time
+// to elapse before they are popped from the 'waiting' queue and added to the 'active' queue.
 type DelayingInterface interface {
 	Interface
 	// AddAfter adds an item to the workqueue after the indicated duration has passed
 	AddAfter(item interface{}, duration time.Duration)
 }
 
+// ExtendedDelayingInterface is a delaying queue with added methods allowing finer control
+// over the delaying queue behaviours and fully asynchronous calls.
+type ExtendedDelayingInterface interface {
```

Comment:
PTAL also at these resolved answers 
- https://github.com/kubernetes/kubernetes/pull/112328#discussion_r1154767849
- https://github.com/kubernetes/kubernetes/pull/112328#discussion_r1154777396

Reply:
Having made the inspectable interface private, we are left with the original DelayingInterface and the new ExpandedDelayingInterface.  The idea behind the new interface is so that clients of the old DelayingInterface will not have to add new methods and will receive exactly the same behaviour as they used to.

I have backtracked on making any changes to the default behaviour - instead we will need to move to the new interface if/when we want to benefit from the new options and done waiting.

Reply:
I find it a bit unfortunate that people need to cast the queue any time they want to access the advanced methods. It is hard to understand from the code what you are getting and people might not be aware they need to cast it and could miss out on the features.

I personally do not think it is a big issue to make such a change in client-go as we are just adding options, and there have been examples of this in the past.

It is a short term pain for us and anybody who is importing the client-go to implement the methods, but a long term gain for anybody using these in the future.